crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -865,11 +871,48 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = for_statement;
 
         self.infer_expression(iter);
-        self.infer_expression(target);
+        // TODO more complex assignment targets
+        if let ast::Expr::Name(name) = &**target {
+            self.infer_definition(name);
+        } else {
+            self.infer_expression(target);
+        }
         self.infer_body(body);
         self.infer_body(orelse);
     }
 
+    fn infer_for_statement_definition(
+        &mut self,
+        target: &ast::ExprName,
+        iterable: &ast::Expr,
+        definition: Definition<'db>,
+    ) {
+        let expression = self.index.expression(iterable);
+        let result = infer_expression_types(self.db, expression);
+        self.extend(result);
+        let iterable_ty = self
+            .types
+            .expression_ty(iterable.scoped_ast_id(self.db, self.scope));
+
+        // TODO(Alex): only a valid iterable if the *type* of `iterable_ty` has an `__iter__`
+        // member (dunders are never looked up on an instance)
+        let _dunder_iter_ty = iterable_ty.member(self.db, &ast::name::Name::from("__iter__"));
+
+        // TODO(Alex):
+        // - infer the return type of the `__iter__` method, which gives us the iterator
+        // - lookup the `__next__` method on the iterator
+        // - infer the return type of the iterator's `__next__` method,
+        //   which gives us the type of the variable being bound here
+        //   (...or the type of the object being unpacked into multiple definitions, if it's something like
+        //   `for k, v in d.items(): ...`)
+        let loop_var_value_ty = Type::Unknown;
```

Comment:
I think this is probably not really the right place for this TODO? Since the inferring of the iterator type (and the unpackingn of whatever is returned from the iterator's `__next__` method) really needs to be done at a higher level? but I'm also not really sure where else to put this TODO?

Reply:
In the current architecture, this is the right place for the TODO; we will have to repeat this logic for each definition, in the case of unpacking. We don't currently have another layer available at which to cache it. Caching also has an overhead and sometimes it's better to just repeat some work. But we can also re-evaluate this when we implement unpacking.

Reply:
I'm not 100% sure we're on the same page here (but this may also just be me still getting used to this architecture) — anyway, I don't think it's massively important where this TODO lives, so I'll leave this for now!